// @patternfly/react-topology/dist/js/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GRAPH_POSITION_CHANGE_EVENT = exports.GRAPH_LAYOUT_END_EVENT = exports.NODE_POSITIONED_EVENT = exports.NODE_COLLAPSE_CHANGE_EVENT = exports.REMOVE_CHILD_EVENT = exports.ELEMENT_VISIBILITY_CHANGE_EVENT = exports.ADD_CHILD_EVENT = exports.isEdge = exports.isNode = exports.isGraph = exports.ModelKind = exports.NodeShape = exports.AnchorEnd = void 0;
var AnchorEnd;
(function (AnchorEnd) {
    AnchorEnd[AnchorEnd["target"] = 0] = "target";
    AnchorEnd[AnchorEnd["source"] = 1] = "source";
    AnchorEnd[AnchorEnd["both"] = 2] = "both";
})(AnchorEnd = exports.AnchorEnd || (exports.AnchorEnd = {}));
var NodeShape;
(function (NodeShape) {
    NodeShape[NodeShape["circle"] = 0] = "circle";
    NodeShape[NodeShape["rect"] = 1] = "rect";
})(NodeShape = exports.NodeShape || (exports.NodeShape = {}));
var ModelKind;
(function (ModelKind) {
    ModelKind["graph"] = "graph";
    ModelKind["node"] = "node";
    ModelKind["edge"] = "edge";
})(ModelKind = exports.ModelKind || (exports.ModelKind = {}));
const isGraph = (element) => element && element.getKind() === ModelKind.graph;
exports.isGraph = isGraph;
const isNode = (element) => element && element.getKind() === ModelKind.node;
exports.isNode = isNode;
const isEdge = (element) => element && element.getKind() === ModelKind.edge;
exports.isEdge = isEdge;
exports.ADD_CHILD_EVENT = 'element-add-child';
exports.ELEMENT_VISIBILITY_CHANGE_EVENT = 'element-visibility-change';
exports.REMOVE_CHILD_EVENT = 'element-remove-child';
exports.NODE_COLLAPSE_CHANGE_EVENT = 'node-collapse-change';
exports.NODE_POSITIONED_EVENT = 'node-positioned';
exports.GRAPH_LAYOUT_END_EVENT = 'graph-layout-end';
exports.GRAPH_POSITION_CHANGE_EVENT = 'graph-position-change';
